partners_api/taxi_engine.cpp

```diff
@@ -0,0 +1,203 @@
+  return std::find_if(first, last,
+                      [type](IterValueType const & item) { return item.m_type == type; });
+}
+}  // namespace
+
+namespace taxi
+{
+void ResultMaker::Reset(uint64_t requestId, size_t requestsCount,
+                        SuccessfullCallback const & successCallback,
+                        ErrorCallback const & errorCallback)
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+
+  m_requestId = requestId;
+  m_requestsCount = static_cast<uint8_t>(requestsCount);
+  m_successCallback = successCallback;
+  m_errorCallback = errorCallback;
+  m_providers.clear();
+  m_errors.clear();
+}
+
+void ResultMaker::DecrementRequestCount(uint64_t requestId)
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+
+  if (m_requestId != requestId)
+    return;
+
+  DecrementRequestCount();
+}
+
+void ResultMaker::ProcessProducts(uint64_t requestId, Provider::Type type,
+                                  std::vector<Product> const & products)
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+
+  if (m_requestId != requestId)
+    return;
+
+  m_providers.emplace_back(type, products);
+
+  DecrementRequestCount();
+}
+
+void ResultMaker::ProcessError(uint64_t requestId, Provider::Type type, ErrorCode code)
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+
+  if (m_requestId != requestId)
+    return;
+
+  m_errors.emplace_back(type, code);
+
+  DecrementRequestCount();
+}
+
+void ResultMaker::MakeResult(uint64_t requestId) const
+{
+  SuccessfullCallback successCallback;
+  ErrorCallback errorCallback;
+  ProvidersContainer providers;
+  ErrorsContainer errors;
+
+  {
+    std::lock_guard<std::mutex> lock(m_mutex);
+
+    if (m_requestId != requestId || m_requestsCount != 0)
+      return;
+
+    successCallback = m_successCallback;
+    errorCallback = m_errorCallback;
+    providers = m_providers;
+    errors = m_errors;
+  }
+
+  if (providers.empty())
+    return errorCallback(errors, requestId);
+
+  return successCallback(providers, requestId);
+}
+
+void ResultMaker::DecrementRequestCount()
+{
+  CHECK_GREATER(m_requestsCount, 0, ());
+  --m_requestsCount;
+}
+
+Engine::Engine()
+{
+  m_enabledCountries = {{Provider::Type::Yandex, {"Russian Federation"}}};
+  m_disabledCountries = {{Provider::Type::Uber, {"Russian Federation"}}};
+
+  m_apis.emplace_back(Provider::Type::Yandex, my::make_unique<yandex::Api>());
+  m_apis.emplace_back(Provider::Type::Uber, my::make_unique<uber::Api>());
+}
+
+/// Requests list of available products. Returns request identificator immediately.
+uint64_t Engine::GetAvailableProducts(ms::LatLon const & from, ms::LatLon const & to,
+                                      storage::TCountriesVec const & countryIds,
+                                      SuccessfullCallback const & successFn,
+                                      ErrorCallback const & failFn)
+{
+  auto const reqId = ++m_requestId;
+  auto const maker = m_maker;
+
+  maker->Reset(reqId, m_apis.size(), successFn, failFn);
```

Comment:
For me maker seems like a quite light-weight class, why do we need to share a single instance of it and protect it via mutex? Why not to create an instance of maker for each request, and share ownership between callbacks? You'll be able to get rid of mutex.